jni/src/common/Foreign/JNI/Unsafe.hs

```diff
@@ -1078,44 +1049,23 @@ getDirectBufferCapacity (upcast -> jbuffer) = do
     else
       throwIO DirectBufferFailed
 
-data GlobalRefCleaner = GlobalRefCleaner
-  { nextBatchToDelete :: IORef [JObject]
-  , wakeup :: MVar ()
-  , cleanerAsync :: Async ()
-  }
-
-createGlobalRefCleaner :: IO GlobalRefCleaner
-createGlobalRefCleaner = do
-  wakeup <- newEmptyMVar
-  refs <- newIORef []
-  deletingThread <- asyncBound $ runInAttachedThread $ forever $ do
-    takeMVar wakeup
-    xs <- atomicModifyIORef refs $ \xs -> ([], xs)
-    mapM_ deleteGlobalRefNonFinalized xs
-  return (GlobalRefCleaner refs wakeup deletingThread)
-
-stopGlobalRefCleaner :: GlobalRefCleaner -> IO ()
-stopGlobalRefCleaner = uninterruptibleMask_ . cancel . cleanerAsync
-
-cleanGlobalRef :: GlobalRefCleaner -> JObject -> IO ()
-cleanGlobalRef cleaner obj = do
-  atomicModifyIORef (nextBatchToDelete cleaner) $ \xs -> (obj:xs, ())
-  _ <- tryPutMVar (wakeup cleaner) ()
-  return ()
-
-globalRefCleaner :: IORef GlobalRefCleaner
-{-# NOINLINE globalRefCleaner #-}
-globalRefCleaner = unsafePerformIO $ newIORef uninitializedGlobalRefCleaner
-
-uninitializedGlobalRefCleaner :: GlobalRefCleaner
-uninitializedGlobalRefCleaner = error "Cannot delete a reference: a dedicated thread is not initialized"
-
--- | Submit a global, non-finalized reference for deletion.
--- Passes the reference to a thread attached to the jvm, which performs the deletion.
+-- | A background thread for cleaning global references
+finalizerThread :: IORef BackgroundWorker
+{-# NOINLINE finalizerThread #-}
+finalizerThread = unsafePerformIO $ newIORef uninitializedFinalizerThread
+
+uninitializedFinalizerThread :: BackgroundWorker
+uninitializedFinalizerThread =
+  error "Cannot delete a reference: a dedicated thread is not initialized"
```

Comment:
As the `BackgroundWorker` is not longer dedicated to deleting references, I would change the error message to something simpler, like "Background worker is not initialized"